refactor(hydro): introduce the dao layer and improve class design

maugde opened the pull request:
Linked to the [ANT-3084](https://gopro-tickets.rte-france.com/browse/ANT-3084) issue.

Refactoring of the hydro component into a dao design.

NB the in memory implementation perhaps needs improvement.

Commit:
refactor(hydro): add hydro dao and file_study_hydro_dao, update hydro relative files, update in-memory study dao class

Commit:
fix(hydro): add implementation of methods in `in_memory_study_dao.py`

Commit:
chore(hydro): remove TODOs

Commit:
test(hydro): fix tests, path used in get_all_hydro_properties

Commit:
fix(hydro): update `in_memory_study_dao.py`

Commit:
chore(hydro): minor fixes

simplify paths, update in_memory_study_dao and hydro_dao save methods

TheoPascoli commented on antarest/study/business/model/hydro_model.py:
```diff
@@ -19,12 +19,8 @@
-
-
-HYDRO_PATH = ["input", "hydro", "hydro"]
+HYDRO_PATH = "input/hydro/hydro"
```
I think we can set them as list so we don't have to split everytime we want to use them

MartinBelthle replied:
I agree

TheoPascoli reviewed (commented):
Hello Mauranne, it looks good, good job !
I just have a suggestion: in the TestUpdateHydroManagement, some tests have incorrect data type, you could fix those errors (Expected type 'HydroManagementUpdate', got 'dict[str, float | bool]' instead). 
And the UpdateInflowStructure command have 0 test, you could add some.
Also, you can remove the "!" in the PR title as it's not a breaking change.

MartinBelthle commented on antarest/study/business/areas/hydro_management.py:
```diff
@@ -9,16 +9,14 @@
 # SPDX-License-Identifier: MPL-2.0
 #
 # This file is part of the Antares project.
+from typing import Dict
```
Minor but no need for that you can let `dict` instead of `Dict` as it was before

MartinBelthle commented on antarest/study/dao/file/file_study_hydro_dao.py:
```diff
@@ -0,0 +1,75 @@
+    def get_inflow_structure(self, area_id: str) -> InflowStructure:
+        file_study = self.get_file_study()
+        inter_monthly_correlation = file_study.tree.get(INFLOW_PATH.format(area_id=area_id).split("/")).get(
+            "intermonthly-correlation", 0.5
```
I think we shouldn't put this default value here as it is inside the InflowStructure class and shouldn't be knowed elsewhere.

MartinBelthle commented on antarest/study/dao/file/file_study_hydro_dao.py:
```diff
@@ -0,0 +1,75 @@
+        return InflowStructure(inter_monthly_correlation=inter_monthly_correlation)
+
+    @override
+    def save_hydro_management(self, area_id: str, hydro_data: Dict[str, Any]) -> None:
```
The 2 `save` method should take objects as arguments and not dict as we did inside the other DAOs

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_hydro_management.py:
```diff
@@ -49,18 +49,18 @@ def validate_properties_against_version(self) -> "UpdateHydroManagement":
+        hydro_manager = HydroManagementFileData(**study_data.get_file_study().tree.get(HYDRO_PATH.split("/")))
 
-        new_hydro = current_hydro.get_hydro_management(self.area_id, self.study_version).model_copy(
+        new_hydro = hydro_manager.get_hydro_management(self.area_id, study_data.get_version()).model_copy(
```
We shouldn't do model_copy at several places in the code. We should have a method inside hydro_model that does that. Take the `thermal_model` as an example

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_inflow_structure.py:
```diff
@@ -43,16 +43,17 @@ class UpdateInflowStructure(ICommand):
+        path = INFLOW_PATH.format(area_id=self.area_id).split("/")
 
-        current_inflow = InflowStructureFileData(**study_data.tree.get(path))
+        current_inflow = InflowStructureFileData(**file_study.tree.get(path))
```
Also no one should know the classes `xxxFileData`. They are implemntation details. We should introduce a file, `hydro.py` inside `storage/rawstudy/model/config` that contains these classes and parsing and serializing methods to use. See the thermal example

MartinBelthle commented on antarest/study/web/study_data_blueprint.py:
```diff
@@ -316,7 +316,7 @@ def get_hydro_form_values(uuid: str, area_id: str) -> HydroManagement:
         study = study_service.check_study_access(uuid, StudyPermissionType.READ)
         study_interface = study_service.get_study_interface(study)
-        return study_service.hydro_manager.get_hydro_management(study_interface, area_id)
+        return study_service.hydro_manager.get_hydro_for_area(study_interface, area_id)
```
The previous name was better IMO as we return a HydroManagement object

MartinBelthle commented on tests/study/business/areas/test_hydro_management.py:
```diff
@@ -110,16 +110,16 @@ def test_set_field_values(self, hydro_manager: HydroManager, empty_study_880: Fi
             # simulate regular usage by modifying some values
             modified_data = dict(initial_data)
-            modified_data["intra_daily_modulation"] = 5.0
+            modified_data["intra_daily_modulation"] = str(5.0)
```
Why did this change ?

Commit:
fix(hydro): minor fixes

change `get_hydro_for_area`, add update methods hydro and inflow, fix typing, update paths

Commit:
fix(hydro): fix in_memory save methods

Commit:
chore(hydro): fix HydroManagementFileData methods and update commands

MartinBelthle commented on antarest/study/dao/api/hydro_dao.py:
```diff
@@ -0,0 +1,47 @@
+
+class HydroDao(ReadOnlyHydroDao):
+    @abstractmethod
+    def save_hydro_management(self, area_id: str, hydro_data: HydroManagementFileData) -> None:
```
here the argument should be `hydro_management: HydroManagement`. Same for the inflow structure method.
Because as I said in a previous comment these classes are file system implementation and the goal of the DAO is to hide them to introduce a second implem later , a DB implem. So the DAO should only have `hydro_model.py` classes

MartinBelthle commented on antarest/study/dao/file/file_study_hydro_dao.py:
```diff
@@ -0,0 +1,78 @@
+        all_hydro_properties = {}
+
+        study_data = self.get_file_study()
+        hydro_management_file_data = HydroManagementFileData(**study_data.tree.get(HYDRO_PATH))
```
You should use the parsing /serializing methods you introduced inside the `hydro.py` class

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,118 @@
+        return HydroManagement(**args)
+
+    @staticmethod
+    def _get_fields_to_exclude(study_version: StudyVersion) -> set[str]:
```
This does what we want to.
But we want the code of every DAO classes to look a bit the same for maintenance sake.
So you can check what we do for thermal clusters but to sum up:
- We should introduce the methods `check_attribute_coherence` , `initialize_hydro_management` and `validate_hydro_management_against_version` inside `hydro_model.py`
- We use them inside this file (and inside the commands for `validate_against_version`)

MartinBelthle commented on tests/study/business/areas/test_hydro_management.py:
```diff
@@ -115,8 +115,8 @@ def test_set_field_values(self, hydro_manager: HydroManager, empty_study_880: Fi
-            modified_data["intra_daily_modulation"] = 5.0
-            hydro_manager.update_hydro_management(study, HydroManagementUpdate(**modified_data), area)
+            modified_data.update(**{"intra_daily_modulation": 5.0})
+            hydro_manager.update_hydro_management(study, HydroManagementUpdate.model_construct(**modified_data), area)
```
Minor: Why are we using `model_construct` and not `model_validate` here ?

maugde replied:
mb i used it to debug and forgot to change it

MartinBelthle commented on tests/variantstudy/model/command/test_update_hydro_management.py:
```diff
@@ -46,7 +47,7 @@ def test_lifecycle(self, empty_study_880: FileStudy, empty_study_920: FileStudy,
 
             # Update several properties
-            new_properties = {"inter_daily_breakdown": 3.1, "reservoir": True}
+            new_properties = HydroManagementUpdate.model_construct(inter_daily_breakdown=3.1, reservoir=True)
```
Same question for the 2 places in this file

MartinBelthle reviewed (changes requested):
We're going in the right direction

Commit:
chore(hydro): add default methods for models

Commit:
chore(hydro): tidy up code

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,151 @@
+    return excludes
+
+
+def _add_default_values(data: dict[str, Any], study_version: StudyVersion) -> dict[str, Any]:
```
This shouldn't be necessary as we have the method `initialize_hydro_management`

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,151 @@
+    return hydro_file_data.model_dump()
+
+
+def _get_fields_to_exclude(study_version: StudyVersion) -> set[str]:
```
Idem this shouldn't be needed I think now that you introduced the `validate_against_version` field

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_hydro_management.py:
```diff
@@ -45,22 +47,20 @@ class UpdateHydroManagement(ICommand):
-        current_hydro = HydroManagementFileData(**study_data.tree.get(HYDRO_PATH))
+    def _apply_dao(self, study_data: StudyDao, listener: Optional[ICommandListener] = None) -> CommandOutput:
+        file_study = study_data.get_file_study()
+        file_data = file_study.tree.get(HYDRO_PATH)
```
Here we shouldn't introduce `file_study` notions as it should only be inside the FileStudy implem of the DAO for the reasons I already explained

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_inflow_structure.py:
```diff
@@ -43,16 +44,17 @@ class UpdateInflowStructure(ICommand):
+    def _apply_dao(self, study_data: StudyDao, listener: Optional[ICommandListener] = None) -> CommandOutput:
+        file_study = study_data.get_file_study()
         path = get_inflow_path(self.area_id)
+        file_data = file_study.tree.get(path)
```
Same

MartinBelthle commented on tests/variantstudy/model/command/test_update_hydro_management.py:
```diff
@@ -79,12 +80,12 @@ def test_lifecycle(self, empty_study_880: FileStudy, empty_study_920: FileStudy,
 
             # Update properties
-            new_properties = {"overflow_spilled_cost_difference": 1.4}
+            new_properties = HydroManagementUpdate.model_construct(overflow_spilled_cost_difference=1.4)
```
We should use `model_validate`

MartinBelthle commented on antarest/study/dao/file/file_study_hydro_dao.py:
```diff
@@ -0,0 +1,82 @@
+    def save_hydro_management(self, area_id: str, hydro_management: HydroManagement) -> None:
+        file_study = self.get_file_study()
+        file_data = file_study.tree.get(HYDRO_PATH)
+        file_data = serialize_hydro_management(area_id, hydro_management, file_data, file_study.config.version)
```
It would be better to have a serialize method that only takes the `hydro_management` and then update the `file_data` with it because the method does more than what it's supposed to and it's different from the other serializing methods in the other models

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,151 @@
+        return cls.model_validate(inflow_structure.model_dump(exclude={"id"}))
+
+
+def _parse_hydro_management(
```
Don't really like those but I think we indeed need them. 
Should be methods of the `HydroManagementFileData` class though

MartinBelthle reviewed (changes requested):
Several changes needed

Commit:
chore(hydro): fix classes and methods responsibilities

update commands with study dao methods instead of study_data, fix minor errors

MartinBelthle commented on antarest/study/business/model/hydro_model.py:
```diff
@@ -15,18 +15,17 @@
+HYDRO_PATH = ["input", "hydro", "hydro"]
+
 
 def get_inflow_path(area_id: str) -> list[str]:
```
This is only used inside the file_study implem, it belongs there

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,129 @@
+        return cls.model_validate(inflow_structure.model_dump(exclude={"id"}))
+
+
+def _parse_hydro_management(hydro_file_data: HydroManagementFileData, area_id: str) -> dict[str, Any]:
```
You should also put this as a method of the `HydroManagementFileData` class

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,129 @@
+        hydro_file_data = HydroManagementFileData()
+
+        for key, value in hydro_data.items():
+            current_dict = getattr(hydro_file_data, key, {})
```
This is a weird way to do things, the class is a Pydantic model, seems you can just do this inside the `from_model` method:

```python
lower_area_id = area_id.lower()
hydro_data = hydro_management.model_dump(exclude_none=True)
args = {}
for key, value in hydro_data.items():
     args.setdefault(key, {})[lower_area_id] = value
return HydroManagementFileData.model_validate(args)
```

MartinBelthle commented on antarest/study/storage/rawstudy/model/filesystem/config/hydro.py:
```diff
@@ -0,0 +1,129 @@
+
+def _parse_hydro_management(hydro_file_data: HydroManagementFileData, area_id: str) -> dict[str, Any]:
+    lower_area_id = area_id.lower()
+    data = hydro_file_data.model_dump(mode="json").items()
```
Don't need the `.items()` I think

MartinBelthle commented on antarest/study/dao/file/file_study_hydro_dao.py:
```diff
@@ -0,0 +1,85 @@
+        file_study = self.get_file_study()
+        initial_hydro_data = file_study.tree.get(HYDRO_PATH)
+        new_hydro_data = serialize_hydro_management(hydro_management, area_id, file_study.config.version)
+        updated_hydro_data = {
```
This seems a bit weird but if it works, ok

MartinBelthle reviewed (changes requested):
Last round of comments, only minor ones

MartinBelthle committed:
resolve comments

MartinBelthle committed:
Merge branch 'dev' into refactor/ANT-3084-introduce-the-dao-layer-and-improve-class-design
